graphql module not found when using now deployment guide

<!--    Instructions                                -->
<!--                                                -->
<!-- 1. Remove sections/details you do not complete -->
<!-- 2. Add sections/details useful to you          -->

First raised in https://github.com/graphql-nexus/nexus/discussions/676

#### Nexus Report <!-- Copy the result of `$ nexus report` to this section -->

```json
{
  "node": "v12.14.1",
  "nexus": "^0.20.0-next.47",
  "plugins": [],
  "os": {
    "platform": "darwin",
    "release": "19.3.0"
  },
  "otherDependencies": {},
  "devDependencies": {
    "@types/jest": "25.2.1",
    "jest": "25.4.0",
    "jest-environment-node": "^25.4.0",
    "prettier": "2.0.4",
    "ts-jest": "25.4.0"
  },
  "hasAppModule": true,
  "packageManager": "yarn"
}
```

#### Screenshot

![image](https://user-images.githubusercontent.com/284476/79759126-4b926700-82ec-11ea-9158-c01035bf1969.png)


#### Description <!-- If screenshot not obvious enough -->

```
❯ LOG_LEVEL=trace yarn -s nexus build -d now
 296 — nexus execLayout start
   2 — nexus execLayout done
       | execLayout  {
       |               nodeProject: true,
       |               toolProject: true,
       |               toolCurrentlyPresentInNodeModules: true,
       |               runningLocalTool: true,
       |               thisProcessToolBin: {
       |                 name: 'nexus',
       |                 path: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/.bin/nexus',
       |                 dir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/.bin',
       |                 realPath: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/dist/cli/main.js',
       |                 realDir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/dist/cli'
       |               },
       |               project: {
       |                 dir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world',
       |                 binDir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/.bin',
       |                 nodeModulesDir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules',
       |                 toolBinPath: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/.bin/nexus',
       |                 toolBinRealPath: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/dist/cli/main.js'
       |               }
       |             }
   5 — nexus:layout starting scan
   1 — nexus:layout looking for app module
 174 — nexus:layout done looking for app module
 182 ▲ nexus:tsconfig Please set your tsconfig.json compilerOptions.rootDir to "."
   0 ▲ nexus:tsconfig Please set your tsconfig.json include to have "."
   6 — nexus:tsconfig finished read
   1 — nexus:layout completed scan
       | result  {
       |           app: { exists: true, path: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/api/app.ts' },
       |           projectRoot: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world',
       |           sourceRoot: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world',
       |           schemaModules: [ '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/api/graphql/user.ts' ],
       |           project: { name: 'sans-prisma', isAnonymous: false },
       |           tsConfigJson: {
       |             options: {
       |               strict: true,
       |               esModuleInterop: true,
       |               rootDir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world',
       |               outDir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/.build',
       |               configFilePath: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/tsconfig.json'
       |             },
       |             watchOptions: undefined,
       |             fileNames: [
       |               '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/api/app.ts',
       |               '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/api/graphql/user.ts',
       |               '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/tests/environment.ts',
       |               '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/tests/integration.test.ts'
       |             ],
       |             projectReferences: undefined,
       |             typeAcquisition: { enable: false, include: [], exclude: [] },
       |             raw: {
       |               compilerOptions: { strict: true, esModuleInterop: true, rootDir: '.', outDir: 'node_modules/.build' },
       |               include: [ '.' ],
       |               compileOnSave: false
       |             },
       |             errors: [],
       |             wildcardDirectories: { '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world': 1 },
       |             compileOnSave: false,
       |             configFileSpecs: {
       |               filesSpecs: undefined,
       |               includeSpecs: [ '.' ],
       |               excludeSpecs: [ 'node_modules/.build' ],
       |               validatedIncludeSpecs: [ '.' ],
       |               validatedExcludeSpecs: [ 'node_modules/.build' ],
       |               wildcardDirectories: { '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world': 1 }
       |             }
       |           },
       |           packageManagerType: 'yarn',
       |           packageJson: {
       |             content: {
       |               name: 'sans-prisma',
       |               version: '1.0.0',
       |               license: 'MIT',
       |               dependencies: { nexus: '^0.20.0-next.47' },
       |               devDependencies: { '@types/jest': '25.2.1', jest: '25.4.0', 'jest-environment-node': '^25.4.0', prettier: '2.0.4', 'ts-jest': '25.4.0' },
       |               scripts: {
       |                 clean: 'rm -rf dist',
       |                 start: 'node node_modules/.build',
       |                 build: 'nexus build',
       |                 generate: 'nexus generate',
       |                 dev: 'nexus dev',
       |                 test: 'jest',
       |                 doctor: 'nexus doctor',
       |                 'link:nexus': 'yarn -s link nexus && chmod +x node_modules/.bin/nexus'
       |               },
       |               prettier: { semi: false },
       |               jest: { preset: 'ts-jest', testEnvironment: './tests/environment' }
       |             },
       |             path: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/package.json',
       |             dir: '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world'
       |           }
       |         }
   5 — nexus:start-module created start module
       | content  '// GENERATED NEXUS START MODULE\n' +
       |            '\n' +
       |            '\n' +
       |            '\n' +
       |            "process.env.NEXUS_SHOULD_GENERATE_ARTIFACTS = 'true'process.env.NEXUS_DISABLE_SERVER = 'true'\n" +
       |            '\n' +
       |            '\n' +
       |            "process.env.NEXUS_STAGE = 'dev'\n" +
       |            '\n' +
       |            '\n' +
       |            '// Run framework initialization side-effects\n' +
       |            '// Also, import the app for later use\n' +
       |            'const app = require("/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/dist/index.js").default\n' +
       |            '\n' +
       |            '\n' +
       |            '// Last resort error handling\n' +
       |            "process.once('uncaughtException', error => {\n" +
       |            "  app.log.fatal('uncaughtException', { error: error })\n" +
       |            '  process.exit(1)\n' +
       |            '})\n' +
       |            '\n' +
       |            "process.once('unhandledRejection', error => {\n" +
       |            "  app.log.fatal('unhandledRejection', { error: error })\n" +
       |            '  process.exit(1)\n' +
       |            '})\n' +
       |            '\n' +
       |            '\n' +
       |            '// package.json is needed for plugin auto-import system.\n' +
       |            '// On the Zeit Now platform, builds and dev copy source into\n' +
       |            '// new directory. Copying follows paths found in source. Give one here\n' +
       |            '// to package.json to make sure Zeit Now brings it along.\n' +
       |            "require('/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/package.json')\n" +
       |            '\n' +
       |            '\n' +
       |            "// Import the user's schema modules\n" +
       |            '        \n' +
       |            "import '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/api/graphql/user'\n" +
       |            '\n' +
       |            '\n' +
       |            "// Import the user's app module\n" +
       |            'require("/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/api/app")\n' +
       |            '\n' +
       |            '\n' +
       |            '// Boot the server if the user did not already.\n' +
       |            'if (app.__state.isWasServerStartCalled === false) {\n' +
       |            '  app.server.start()\n' +
       |            "}process.env.NEXUS_DISABLE_SERVER = 'false'"
  73 — nexus:plugin loaded plugin entrypoints  --  validPlugins: []
   1 — nexus:tsc Create TypeScript program
3907 — nexus:build Compiling a development build for typegen
  73 — nexus:start-module created start module
       | content  '// GENERATED NEXUS START MODULE\n' +
       |            '\n' +
       |            '\n' +
       |            '\n' +
       |            "process.env.NEXUS_SHOULD_GENERATE_ARTIFACTS = 'true'\n" +
       |            '\n' +
       |            '\n' +
       |            '// Run framework initialization side-effects\n' +
       |            '// Also, import the app for later use\n' +
       |            'const app = require("nexus").default\n' +
       |            '\n' +
       |            '\n' +
       |            '// Last resort error handling\n' +
       |            "process.once('uncaughtException', error => {\n" +
       |            "  app.log.fatal('uncaughtException', { error: error })\n" +
       |            '  process.exit(1)\n' +
       |            '})\n' +
       |            '\n' +
       |            "process.once('unhandledRejection', error => {\n" +
       |            "  app.log.fatal('unhandledRejection', { error: error })\n" +
       |            '  process.exit(1)\n' +
       |            '})\n' +
       |            '\n' +
       |            '\n' +
       |            '// package.json is needed for plugin auto-import system.\n' +
       |            '// On the Zeit Now platform, builds and dev copy source into\n' +
       |            '// new directory. Copying follows paths found in source. Give one here\n' +
       |            '// to package.json to make sure Zeit Now brings it along.\n' +
       |            "require('../package.json')\n" +
       |            '\n' +
       |            '\n' +
       |            "// Import the user's schema modules\n" +
       |            '        \n' +
       |            "import './api/graphql/user'\n" +
       |            '\n' +
       |            '\n' +
       |            "// Import the user's app module\n" +
       |            'require("./api/app")\n' +
       |            '\n' +
       |            '\n' +
       |            '// Boot the server if the user did not already.\n' +
       |            'if (app.__state.isWasServerStartCalled === false) {\n' +
       |            '  app.server.start()\n' +
       |            '}'
   0 — nexus:start-module Transpiling start module
   7 — nexus:build Writing start module to disk
   1 ● nexus:build Running typegen & extracting types from addToContext calls
   1 — nexus:add-to-context-extractor got app source files  --  count: 2
   1 — nexus:add-to-context-extractor found call  --  text: 'addToContext'
   8 — nexus:add-to-context-extractor found call arg
       | text  '(req) => {\n' +
       |         '  return {\n' +
       |         '    req,\n' +
       |         '    db: {\n' +
       |         '      users: {\n' +
       |         '        newton: {\n' +
       |         '          id: "1",\n' +
       |         '          birthyear: "1649",\n' +
       |         '          name: "Newton",\n' +
       |         '        },\n' +
       |         '      },\n' +
       |         '    },\n' +
       |         '  }\n' +
       |         '}'
   1 — nexus:add-to-context-extractor processing prop  --  name: 'req'
  50 — nexus:add-to-context-extractor found prop type alias symbol?  --  found: false
  10 — nexus:add-to-context-extractor found prop type symbol?  --  found: true
   0 — nexus:add-to-context-extractor found name?  --  name: 'Request'
   1 — nexus:add-to-context-extractor processing prop  --  name: 'db'
   1 — nexus:add-to-context-extractor found prop type alias symbol?  --  found: false
   0 — nexus:add-to-context-extractor found prop type symbol?  --  found: true
   1 — nexus:add-to-context-extractor finished compiler extension processing
       | contextTypeContributions  {
       |                             typeImports: [],
       |                             types: [
       |                               '{ req: Request; db: { users: { newton: { id: string; birthyear: string; name: string; }; }; }; }'
       |                             ]
       |                           }
   1 — nexus:typegen start
 299 ✕ app uncaughtException
       | error  Error: Cannot find module './api/graphql/user'
       |        Require stack:
       |        - /Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/dist/index.js
       |            at Function.Module._resolveFilename (internal/modules/cjs/loader.js:793:17)
       |            at Function.Module._load (internal/modules/cjs/loader.js:686:27)
       |            at Module.require (internal/modules/cjs/loader.js:848:19)
       |            at require (internal/modules/cjs/helpers.js:74:18)
       |            at Object.<anonymous> (/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/dist/index.js:23:1)
       |            at Module._compile (internal/modules/cjs/loader.js:955:30)
       |            at Object.Module._extensions..js (internal/modules/cjs/loader.js:991:10)
       |            at Module.load (internal/modules/cjs/loader.js:811:32)
       |            at Function.Module._load (internal/modules/cjs/loader.js:723:14)
       |            at Function.Module.runMain (internal/modules/cjs/loader.js:1043:10) {
       |          code: 'MODULE_NOT_FOUND',
       |          requireStack: [ '/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/dist/index.js' ]
       |        }
 412 ✕ nexus:build failed to generate artifacts
       | error  Error: 
       |              Nexus artifact generation failed with exit code "1". The following stderr was captured:
       |        
       |                  null
       |            
       |            at Object.generateArtifacts (/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/src/lib/build/artifact-generation.ts:30:11)
       |            at Object.buildNexusApp (/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/src/lib/build/build.ts:85:5)
       |            at Build.parse (/Users/jasonkuhrt/projects/graphql-nexus/examples/hello-world/node_modules/nexus/src/cli/commands/build.ts:32:5)
```

#### Repro <!-- Steps or repo link -->

1. Go to hello world example
1. `yarn -s build -d now`

--- a/src/lib/build/build.ts
+++ b/src/lib/build/build.ts
@@ -1,5 +1,6 @@
 import { stripIndent } from 'common-tags'
 import * as FS from 'fs-jetpack'
+import * as Path from 'path'
 import * as Layout from '../../lib/layout'
 import { compile, createTSProgram, deleteTSIncrementalFile } from '../../lib/tsc'
 import {
@@ -33,9 +34,10 @@ export async function buildNexusApp(settings: BuildSettings) {
 
   const startTime = Date.now()
   const deploymentTarget = normalizeTarget(settings.target)
+  const buildOutput = settings.output ?? computeBuildOutputFromTarget(deploymentTarget) ?? undefined
 
   const layout = await Layout.create({
-    buildOutputRelative: settings.output ?? computeBuildOutputFromTarget(deploymentTarget) ?? undefined,
+    buildOutput,
     entrypointPath: settings.entrypoint,
   })
 
@@ -70,9 +72,9 @@ export async function buildNexusApp(settings: BuildSettings) {
     startModule: prepareStartModule(
       tsBuilder,
       createStartModuleContent({
+        layout,
         plugins: manifests,
         internalStage: 'build',
-        layout: layout,
       })
     ),
   })
@@ -113,7 +115,7 @@ export async function buildNexusApp(settings: BuildSettings) {
   })
 
   log.info('success', {
-    buildOutput: layout.buildOutputRelative,
+    buildOutput: Path.relative(layout.projectRoot, layout.buildOutput),
     time: Date.now() - startTime,
   })
 

--- a/src/lib/build/deploy-target.ts
+++ b/src/lib/build/deploy-target.ts
@@ -2,13 +2,14 @@ import chalk from 'chalk'
 import { stripIndent } from 'common-tags'
 import * as fs from 'fs-jetpack'
 import * as path from 'path'
+import * as Path from 'path'
 import { DEFAULT_BUILD_FOLDER_PATH_RELATIVE_TO_PROJECT_ROOT, Layout } from '../../lib/layout'
 import { START_MODULE_NAME } from '../../runtime/start/start-module'
 import { findFileRecurisvelyUpwardSync } from '../fs'
 import { rootLogger } from '../nexus-logger'
 import { fatal } from '../process'
 
-const log = rootLogger.child(__filename)
+const log = rootLogger.child('build')
 
 /**
  * If you add a new deploy target, please start by adding a new item to the `SUPPORTED_DEPLOY_TARGETS`
@@ -75,7 +76,7 @@ interface NowJson {
  */
 function validateNow(layout: Layout): ValidatorResult {
   const maybeNowJson = findFileRecurisvelyUpwardSync('now.json', { cwd: layout.projectRoot })
-  const startModulePath = `${layout.buildOutputRelative}/${START_MODULE_NAME}.js`
+  const startModulePath = `${layout.buildOutput}/${START_MODULE_NAME}.js`
   let isValid = true
 
   // Make sure there's a now.json file
@@ -105,7 +106,9 @@ function validateNow(layout: Layout): ValidatorResult {
     // Make sure the now.json file has the right `builds` values
     if (
       !nowJson.builds ||
-      !nowJson.builds.find((build) => build.src === startModulePath && build.use === '@now/node')
+      !nowJson.builds.find(
+        (build) => Path.join(maybeNowJson.dir, build.src) === startModulePath && build.use === '@now/node'
+      )
     ) {
       log.error(`We could not find a proper builder in your \`now.json\` file`)
       log.error(`Found: "builds": ${JSON.stringify(nowJson.builds)}`)
@@ -123,7 +126,7 @@ function validateNow(layout: Layout): ValidatorResult {
     }
 
     // Make sure the now.json file has the right `routes` values
-    if (!nowJson.routes?.find((route) => route.dest === startModulePath)) {
+    if (!nowJson.routes?.find((route) => Path.join(maybeNowJson.dir, route.dest) === startModulePath)) {
       log.error(`We could not find a route property that redirects to your api in your \`now.json\` file.`)
       log.error(`Found: "routes": ${JSON.stringify(nowJson.routes)}`)
       log.error(`Expected: "routes": [{ src: '/.*', dest: "${startModulePath}" }, ...]`)
@@ -192,17 +195,17 @@ function validateHeroku(layout: Layout): ValidatorResult {
     // Make sure there's a start script
     if (!pcfg.scripts?.start) {
       log.error(
-        `Please add the following to your \`package.json\` file: "scripts": { "start": "node ${layout.buildOutputRelative}" }`
+        `Please add the following to your \`package.json\` file: "scripts": { "start": "node ${layout.buildOutput}" }`
       )
       console.log()
       isValid = false
     }
 
     // Make sure the start script starts the built server
-    if (!pcfg.scripts?.start?.includes(`node ${layout.buildOutputRelative}`)) {
+    if (!pcfg.scripts?.start?.includes(`node ${layout.buildOutput}`)) {
       log.error(`Please make sure your \`start\` script points to your built server`)
       log.error(`Found: "${pcfg.scripts?.start}"`)
-      log.error(`Expected: "node ${layout.buildOutputRelative}"`)
+      log.error(`Expected: "node ${layout.buildOutput}"`)
       console.log()
       isValid = false
     }

--- a/src/lib/build/deploy-target.ts
+++ b/src/lib/build/deploy-target.ts
@@ -1,7 +1,6 @@
 import chalk from 'chalk'
 import { stripIndent } from 'common-tags'
 import * as fs from 'fs-jetpack'
-import * as path from 'path'
 import * as Path from 'path'
 import { DEFAULT_BUILD_FOLDER_PATH_RELATIVE_TO_PROJECT_ROOT, Layout } from '../../lib/layout'
 import { START_MODULE_NAME } from '../../runtime/start/start-module'
@@ -97,7 +96,7 @@ function validateNow(layout: Layout): ValidatorResult {
         "routes": [{ "src": "/.*", "dest": "${startModulePath}" }]
       }
     `
-    const nowJsonPath = path.join(layout.projectRoot, 'now.json')
+    const nowJsonPath = Path.join(layout.projectRoot, 'now.json')
     fs.write(nowJsonPath, nowJsonContent)
     log.warn(`No \`now.json\` file were found. We scaffolded one for you in ${nowJsonPath}`)
   } else {
@@ -192,20 +191,22 @@ function validateHeroku(layout: Layout): ValidatorResult {
       isValid = false
     }
 
+    const relativeBuildOutput = Path.relative(layout.packageJson.dir, layout.buildOutput)
+
     // Make sure there's a start script
     if (!pcfg.scripts?.start) {
       log.error(
-        `Please add the following to your \`package.json\` file: "scripts": { "start": "node ${layout.buildOutput}" }`
+        `Please add the following to your \`package.json\` file: "scripts": { "start": "node ${relativeBuildOutput}" }`
       )
       console.log()
       isValid = false
     }
 
     // Make sure the start script starts the built server
-    if (!pcfg.scripts?.start?.includes(`node ${layout.buildOutput}`)) {
+    if (!pcfg.scripts?.start?.includes(`node ${relativeBuildOutput}`)) {
       log.error(`Please make sure your \`start\` script points to your built server`)
       log.error(`Found: "${pcfg.scripts?.start}"`)
-      log.error(`Expected: "node ${layout.buildOutput}"`)
+      log.error(`Expected to contain: "node ${relativeBuildOutput}"`)
       console.log()
       isValid = false
     }

--- a/src/lib/layout/index.spec.ts
+++ b/src/lib/layout/index.spec.ts
@@ -41,8 +41,10 @@ function tsconfigContent(input: TsConfigJson): string {
   return JSON.stringify(input)
 }
 
+const defaultTsConfigContent = { compilerOptions: { rootDir: '.' }, include: ['.'] }
+
 const fsTsConfig = {
-  'tsconfig.json': tsconfigContent({ compilerOptions: { rootDir: '.' }, include: ['.'] }),
+  'tsconfig.json': tsconfigContent(defaultTsConfigContent),
 }
 
 const layoutContext = TestContext.create((input: TestContext.TmpDirContribution) => {
@@ -50,8 +52,12 @@ const layoutContext = TestContext.create((input: TestContext.TmpDirContribution)
     setup(spec: FSSpec = {}) {
       writeFSSpec(input.tmpDir, spec)
     },
-    async scan(opts?: { entrypointPath?: string }) {
-      const data = await Layout.create({ cwd: input.tmpDir, entrypointPath: opts?.entrypointPath })
+    async scan(opts?: { entrypointPath?: string; buildOutput?: string }) {
+      const data = await Layout.create({
+        cwd: input.tmpDir,
+        entrypointPath: opts?.entrypointPath,
+        buildOutput: opts?.buildOutput,
+      })
       mockedStdoutBuffer = mockedStdoutBuffer.split(input.tmpDir).join('__DYNAMIC__')
       return repalceInObject(input.tmpDir, '__DYNAMIC__', data.data)
     },
@@ -129,8 +135,8 @@ describe('tsconfig', () => {
       ▲ nexus:tsconfig Please set your tsconfig.json include to have \\".\\"
       "
     `)
-    expect(layout.tsConfigJson.raw.compilerOptions.rootDir).toEqual('.')
-    expect(layout.tsConfigJson.raw.include).toEqual(['.'])
+    expect(layout.tsConfig.content.raw.compilerOptions.rootDir).toEqual('.')
+    expect(layout.tsConfig.content.raw.include).toEqual(['.'])
   })
 
   it('will fatal message and exit if error reading file', async () => {
@@ -226,88 +232,15 @@ it('finds nested graphql modules', async () => {
 
   const result = await ctx.scan()
 
-  expect(result).toMatchInlineSnapshot(`
-    Object {
-      "app": Object {
-        "exists": true,
-        "path": "__DYNAMIC__/src/app.ts",
-      },
-      "buildOutputRelative": "node_modules/.build",
-      "packageJson": Object {
-        "dir": "__DYNAMIC__",
-        "path": "__DYNAMIC__/package.json",
-      },
-      "packageManagerType": "npm",
-      "project": Object {
-        "isAnonymous": true,
-        "name": "anonymous",
-      },
-      "projectRoot": "__DYNAMIC__",
-      "schemaModules": Array [
-        "__DYNAMIC__/src/graphql/1.ts",
-        "__DYNAMIC__/src/graphql/2.ts",
-        "__DYNAMIC__/src/graphql/graphql/3.ts",
-        "__DYNAMIC__/src/graphql/graphql/4.ts",
-        "__DYNAMIC__/src/graphql/graphql/graphql/5.ts",
-        "__DYNAMIC__/src/graphql/graphql/graphql/6.ts",
-      ],
-      "sourceRoot": "__DYNAMIC__",
-      "startModuleInPath": "__DYNAMIC__/index.ts",
-      "startModuleOutPath": "__DYNAMIC__/node_modules/.build/index.js",
-      "tsConfigJson": Object {
-        "compileOnSave": false,
-        "configFileSpecs": Object {
-          "excludeSpecs": Array [
-            "node_modules/.build",
-          ],
-          "includeSpecs": Array [
-            ".",
-          ],
-          "validatedExcludeSpecs": Array [
-            "node_modules/.build",
-          ],
-          "validatedIncludeSpecs": Array [
-            ".",
-          ],
-          "wildcardDirectories": Object {
-            "__DYNAMIC__": 1,
-          },
-        },
-        "errors": Array [],
-        "fileNames": Array [
-          "__DYNAMIC__/src/app.ts",
-          "__DYNAMIC__/src/graphql/1.ts",
-          "__DYNAMIC__/src/graphql/2.ts",
-          "__DYNAMIC__/src/graphql/graphql/3.ts",
-          "__DYNAMIC__/src/graphql/graphql/4.ts",
-          "__DYNAMIC__/src/graphql/graphql/graphql/5.ts",
-          "__DYNAMIC__/src/graphql/graphql/graphql/6.ts",
-        ],
-        "options": Object {
-          "configFilePath": "__DYNAMIC__/tsconfig.json",
-          "outDir": "__DYNAMIC__/node_modules/.build",
-          "rootDir": "__DYNAMIC__",
-        },
-        "raw": Object {
-          "compileOnSave": false,
-          "compilerOptions": Object {
-            "outDir": "node_modules/.build",
-            "rootDir": ".",
-          },
-          "include": Array [
-            ".",
-          ],
-        },
-        "typeAcquisition": Object {
-          "enable": false,
-          "exclude": Array [],
-          "include": Array [],
-        },
-        "wildcardDirectories": Object {
-          "__DYNAMIC__": 1,
-        },
-      },
-    }
+  expect(result.schemaModules).toMatchInlineSnapshot(`
+    Array [
+      "__DYNAMIC__/src/graphql/1.ts",
+      "__DYNAMIC__/src/graphql/2.ts",
+      "__DYNAMIC__/src/graphql/graphql/3.ts",
+      "__DYNAMIC__/src/graphql/graphql/4.ts",
+      "__DYNAMIC__/src/graphql/graphql/graphql/5.ts",
+      "__DYNAMIC__/src/graphql/graphql/graphql/6.ts",
+    ]
   `)
 })
 
@@ -423,6 +356,76 @@ it('does not take custom entrypoint as schema module if its inside a graphql/ fo
       ],
     }
   `)
+})
+
+describe('build output', () => {
+  it(`defaults to node_modules/.build`, async () => {
+    await ctx.setup({ ...fsTsConfig, 'graphql.ts': '' })
+    const result = await ctx.scan()
+
+    expect({
+      buildOutput: result.buildOutput,
+      startModuleInPath: result.startModuleInPath,
+      startModuleOutPath: result.startModuleOutPath,
+    }).toMatchInlineSnapshot(`
+      Object {
+        "buildOutput": "__DYNAMIC__/node_modules/.build",
+        "startModuleInPath": "__DYNAMIC__/index.ts",
+        "startModuleOutPath": "__DYNAMIC__/node_modules/.build/index.js",
+      }
+    `)
+  })
+
+  it(`use tsconfig.json outDir is no custom output is used`, async () => {
+    await ctx.setup({
+      'tsconfig.json': tsconfigContent({
+        ...defaultTsConfigContent,
+        compilerOptions: {
+          ...defaultTsConfigContent.compilerOptions,
+          outDir: 'dist',
+        },
+      }),
+      'graphql.ts': '',
+    })
+    const result = await ctx.scan()
+
+    expect({
+      buildOutput: result.buildOutput,
+      startModuleInPath: result.startModuleInPath,
+      startModuleOutPath: result.startModuleOutPath,
+    }).toMatchInlineSnapshot(`
+      Object {
+        "buildOutput": "__DYNAMIC__/dist",
+        "startModuleInPath": "__DYNAMIC__/index.ts",
+        "startModuleOutPath": "__DYNAMIC__/dist/index.js",
+      }
+    `)
+  })
+  it(`override tsconfig.json outDir is a custom output is used`, async () => {
+    await ctx.setup({
+      'tsconfig.json': tsconfigContent({
+        ...defaultTsConfigContent,
+        compilerOptions: {
+          ...defaultTsConfigContent.compilerOptions,
+          outDir: 'dist',
+        },
+      }),
+      'graphql.ts': '',
+    })
+    const result = await ctx.scan({ buildOutput: 'custom-output' })
+
+    expect({
+      buildOutput: result.buildOutput,
+      startModuleInPath: result.startModuleInPath,
+      startModuleOutPath: result.startModuleOutPath,
+    }).toMatchInlineSnapshot(`
+      Object {
+        "buildOutput": "__DYNAMIC__/custom-output",
+        "startModuleInPath": "__DYNAMIC__/index.ts",
+        "startModuleOutPath": "__DYNAMIC__/custom-output/index.js",
+      }
+    `)
+  })
 })
 
 describe('source root', () => {

--- a/src/lib/layout/layout.ts
+++ b/src/lib/layout/layout.ts
@@ -42,7 +42,10 @@ export type ScanResult = {
   sourceRoot: string
   projectRoot: string
   schemaModules: string[]
-  tsConfigJson: ParsedCommandLine
+  tsConfig: {
+    content: ParsedCommandLine
+    path: string
+  }
   packageManagerType: PackageManager.PackageManager['type']
   packageJson: null | {
     dir: string
@@ -71,7 +74,7 @@ export type ScanResult = {
  * the dynamic scan results.
  */
 export type Data = ScanResult & {
-  buildOutputRelative: string
+  buildOutput: string
   startModuleOutPath: string
   startModuleInPath: string
 }
@@ -97,7 +100,7 @@ interface Options {
   /**
    * The place to output the build, relative to project root.
    */
-  buildOutputRelative?: string
+  buildOutput?: string
   entrypointPath?: string
   cwd?: string
 }
@@ -113,22 +116,20 @@ export async function create(options?: Options): Promise<Layout> {
   const cwd = options?.cwd ?? process.cwd()
   const normalizedEntrypoint = normalizeEntrypoint(options?.entrypointPath, cwd)
   // TODO lodash merge defaults or something
-  const optionsMerged = {
-    buildOutputRelative: options?.buildOutputRelative ?? optionDefaults.buildOutput,
-    cwd,
-    entrypointPath: normalizedEntrypoint,
+
+  const scanResult = await scan({ cwd, entrypointPath: normalizedEntrypoint })
+  const buildOutput = getBuildOutput(options?.buildOutput, scanResult)
+  const outputInfo = {
+    buildOutput,
+    startModuleInPath: Path.join(scanResult.sourceRoot, START_MODULE_NAME + '.ts'),
+    startModuleOutPath: Path.join(buildOutput, START_MODULE_NAME + '.js'),
   }
 
-  const data = await scan({ cwd: optionsMerged.cwd, entrypointPath: optionsMerged.entrypointPath })
+  log.trace('additional layout data', { data: outputInfo })
+
   const layout = createFromData({
-    ...data,
-    buildOutputRelative: optionsMerged.buildOutputRelative,
-    startModuleInPath: Path.join(data.sourceRoot, START_MODULE_NAME + '.ts'),
-    startModuleOutPath: Path.join(
-      data.projectRoot,
-      optionsMerged.buildOutputRelative,
-      START_MODULE_NAME + '.js'
-    ),
+    ...scanResult,
+    ...outputInfo,
   })
 
   /**
@@ -170,13 +171,12 @@ export function createFromData(layoutData: Data): Layout {
  */
 export async function scan(opts?: { cwd?: string; entrypointPath?: string }): Promise<ScanResult> {
   log.trace('starting scan')
-
   const projectRoot = opts?.cwd ?? process.cwd()
   const packageManagerType = await PackageManager.detectProjectPackageManager({ projectRoot })
   const maybePackageJson = findPackageJson({ projectRoot })
   const maybeAppModule = opts?.entrypointPath ?? findAppModule({ projectRoot })
   let maybeSchemaModules = Schema.findDirOrModules({ projectRoot })
-  const tsConfigJson = await readOrScaffoldTsconfig({
+  const tsConfig = await readOrScaffoldTsconfig({
     projectRoot,
   })
 
@@ -190,12 +190,12 @@ export async function scan(opts?: { cwd?: string; entrypointPath?: string }): Pr
       maybeAppModule === null
         ? ({ exists: false, path: maybeAppModule } as const)
         : ({ exists: true, path: maybeAppModule } as const),
-    projectRoot: projectRoot,
-    sourceRoot: tsConfigJson.options.rootDir!,
+    projectRoot,
+    sourceRoot: tsConfig.content.options.rootDir!,
     schemaModules: maybeSchemaModules,
     project: readProjectInfo(opts),
-    tsConfigJson: tsConfigJson,
-    packageManagerType: packageManagerType,
+    tsConfig,
+    packageManagerType,
     packageJson: maybePackageJson,
   }
 
@@ -206,6 +206,7 @@ export async function scan(opts?: { cwd?: string; entrypointPath?: string }): Pr
   }
 
   log.trace('completed scan', { result })
+
   return result
 }
 
@@ -388,32 +389,6 @@ function findPackageJson(opts: { projectRoot: string }): ScanResult['packageJson
   }
 }
 
-function getShallowestPath(paths: string[]) {
-  return paths.sort((a, b) => {
-    return a.split(Path.sep).length - b.split(Path.sep).length
-  })[0]
-}
-
-function getSourceRoot(
-  maybeAppModule: string | null,
-  maybeSchemaModules: string[],
-  opts: { projectRoot: string }
-) {
-  let sourceRoot: string
-
-  if (maybeAppModule) {
-    sourceRoot = Path.dirname(maybeAppModule)
-  } else {
-    if (maybeSchemaModules.length !== 0) {
-      sourceRoot = Path.dirname(getShallowestPath(maybeSchemaModules))
-    } else {
-      sourceRoot = opts.projectRoot
-    }
-  }
-
-  return sourceRoot
-}
-
 function normalizeEntrypoint(entrypoint: string | undefined, cwd: string): string | undefined {
   if (!entrypoint) {
     return undefined
@@ -430,4 +405,18 @@ function normalizeEntrypoint(entrypoint: string | undefined, cwd: string): strin
   }
 
   return absoluteEntrypoint
+}
+
+/**
+ * Get the relative output build path
+ * Precedence: User's input > tsconfig.json's outDir > default
+ */
+function getBuildOutput(buildOutput: string | undefined, scanResult: ScanResult): string {
+  const output = buildOutput ?? scanResult.tsConfig.content.options.outDir ?? optionDefaults.buildOutput
+
+  if (Path.isAbsolute(output)) {
+    return output
+  }
+
+  return Path.join(scanResult.projectRoot, output)
 }

--- a/src/lib/layout/tsconfig.ts
+++ b/src/lib/layout/tsconfig.ts
@@ -16,7 +16,7 @@ const diagnosticHost: ts.FormatDiagnosticsHost = {
 export async function readOrScaffoldTsconfig(input: {
   projectRoot: string
   overrides?: { outRoot?: string }
-}): Promise<ts.ParsedCommandLine> {
+}): Promise<{ content: ts.ParsedCommandLine, path: string }> {
   let tsconfigPath = ts.findConfigFile(input.projectRoot, ts.sys.fileExists, 'tsconfig.json')
 
   if (!tsconfigPath) {
@@ -103,7 +103,7 @@ export async function readOrScaffoldTsconfig(input: {
 
   log.trace('finished read')
 
-  return tsconfig
+  return { content: tsconfig, path: tsconfigPath }
 }
 
 function checkNoTsConfigErrors(tsconfig: ts.ParsedCommandLine) {

--- a/src/lib/tsc.ts
+++ b/src/lib/tsc.ts
@@ -29,7 +29,8 @@ export function createTSProgram(
     rootNames: layout.schemaModules.concat(layout.app.exists ? [layout.app.path] : []),
     options: {
       ...compilerCacheOptions,
-      ...layout.tsConfigJson.options,
+      ...layout.tsConfig.content.options,
+      outDir: layout.buildOutput,
     },
   })
 
@@ -79,7 +80,7 @@ export function compile(
 ): void {
   if (options?.removePreviousBuild === true) {
     log.trace('remove previous build folder if present')
-    fs.remove(layout.buildOutputRelative)
+    fs.remove(layout.buildOutput)
   }
 
   log.trace('emit transpiled modules')

--- a/src/runtime/start/start-module.ts
+++ b/src/runtime/start/start-module.ts
@@ -78,7 +78,7 @@ export function createStartModuleContent(config: StartModuleConfig): string {
       require('${
         config.absoluteModuleImports
           ? config.layout.packageJson.path
-          : Path.relative(config.layout.buildOutputRelative, config.layout.packageJson.path)
+          : Path.relative(config.layout.buildOutput, config.layout.packageJson.path)
       }')
     `
   }

